Modules/_sqlite/module.c

```diff
@@ -351,17 +351,17 @@ do {                                           \
     }                                          \
 } while (0)
 
-#define ADD_EXCEPTION(module, name, exc, base)                  \
-do {                                                            \
-    exc = PyErr_NewException(MODULE_NAME "." name, base, NULL); \
-    if (!exc) {                                                 \
-        goto error;                                             \
-    }                                                           \
-    int res = PyModule_AddObjectRef(module, name, exc);         \
-    Py_DECREF(exc);                                             \
-    if (res < 0) {                                              \
-        goto error;                                             \
-    }                                                           \
+#define ADD_EXCEPTION(module, state, name, base)                         \
+do {                                                                     \
+    state->name = PyErr_NewException(MODULE_NAME "." #name, base, NULL); \
+    if (state->name == NULL) {                                           \
+        goto error;                                                      \
+    }                                                                    \
+    int res = PyModule_AddObjectRef(module, #name, state->name);         \
+    Py_DECREF(state->name);                                              \
+    if (res < 0) {                                                       \
+        goto error;                                                      \
+    }                                                                    \
```

Comment:
Now that I look at it, the refcounting doesn't look right here:

- PyErr_NewException creates a reference
- PyModule_AddObjectRef creates another reference, held by the module dict
- Py_DECREF removes a reference

This means that the only reference to the exception is now held by the module dict, from which it can be removed by the user.
Another reference needs to be held by the module state (which is currently global/immortal, so the reference should currently be leaked).
Am I missing anything?

---

And while you're refactoring, `ADD_TYPE` can be used for exception types as well.

Reply:
> This means that the only reference to the exception is now held by the module dict, from which it can be removed by the user.

Indeed. The `Py_DECREF` must go.

> And while you're refactoring, `ADD_TYPE` can be used for exception types as well.

Yes, thanks!

Reply:
Fixed in 54a412b5a252dd560eda4cf71870d0e4199339cd and b7c49ffeabcfdb95429925b25287dd8ac38da164